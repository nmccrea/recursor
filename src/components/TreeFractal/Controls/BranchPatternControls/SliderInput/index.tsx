import React from "react"
import { useSelector, useDispatch } from "react-redux"
import { BranchPatternInputSelector } from "../../../../../state/branchPatterns/selectors"
import { BranchPatternInputActionCreator } from "../../../../../state/branchPatterns/actions"
import {
  BranchPatternNumericInputs,
  BranchPatternNumericInputKey,
} from "../../../../../state/branchPatterns/types"

export interface SliderInputProps<Key extends BranchPatternNumericInputKey> {
  selector: BranchPatternInputSelector<Key>
  actionCreator: BranchPatternInputActionCreator<Key>
  label: string
  min: BranchPatternNumericInputs[Key]
  max: BranchPatternNumericInputs[Key]
  step: BranchPatternNumericInputs[Key]
}

const SliderInput = <Key extends BranchPatternNumericInputKey>({
  selector,
  actionCreator,
  label,
  min,
  max,
  step,
}: SliderInputProps<Key>) => {
  const value = useSelector(selector)
  const dispatch = useDispatch()

  return (
    <label>
      {label}

      <input
        type="range"
        value={value}
        min={min}
        max={max}
        step={step}
        onChange={(event) =>
          dispatch(actionCreator(Number(event.target.value)))
        }
      />
    </label>
  )
}

export default SliderInput
