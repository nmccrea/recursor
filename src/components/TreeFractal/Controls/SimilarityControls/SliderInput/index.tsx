import React from "react"
import { useSelector, useDispatch } from "react-redux"
import { SimilarityInputSelector } from "../../../../../state/similarities/selectors"
import { SimilarityInputActionCreator } from "../../../../../state/similarities/actions"
import {
  SimilarityNumericInputs,
  SimilarityNumericInputKey,
} from "../../../../../state/similarities/types"

export interface SliderInputProps<Key extends SimilarityNumericInputKey> {
  selector: SimilarityInputSelector<Key>
  actionCreator: SimilarityInputActionCreator<Key>
  label: string
  min: SimilarityNumericInputs[Key]
  max: SimilarityNumericInputs[Key]
  step: SimilarityNumericInputs[Key]
}

const SliderInput = <Key extends SimilarityNumericInputKey>({
  selector,
  actionCreator,
  label,
  min,
  max,
  step,
}: SliderInputProps<Key>) => {
  const value = useSelector(selector)
  const dispatch = useDispatch()

  return (
    <label>
      {label}

      <input
        type="range"
        value={value}
        min={min}
        max={max}
        step={step}
        onChange={(event) =>
          dispatch(actionCreator(Number(event.target.value)))
        }
      />
    </label>
  )
}

export default SliderInput
