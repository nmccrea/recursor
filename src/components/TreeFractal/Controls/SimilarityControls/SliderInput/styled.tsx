import styled from "styled-components"
import { grid } from "../../../../../styles/utilities"

const Container = styled.div`
  display: grid;
  ${grid.columns(3)}
  align-items: stretch;
`

const Label = styled.label`
  ${grid.columnSpan(3)}
  font-size: 0.9em;
`

const Input = styled.input`
  ${grid.columnSpan(2)}
`

const Value = styled.span`
  margin-left: 0.25rem;
  font-size: 0.9em;
  justify-self: flex-end;
  white-space: nowrap;
`

const Min = styled.span`
  justify-self: flex-start;

  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
`

const Max = styled.span`
  justify-self: flex-end;

  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
`

const Unit = styled.span`
  font-size: 0.8em;
  font-style: italic;
  color: gray;
`

export { Container, Label, Input, Value, Min, Max, Unit }
