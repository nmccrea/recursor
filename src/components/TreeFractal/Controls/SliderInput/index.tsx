import React from "react"
import { useSelector, useDispatch } from "react-redux"
import { BranchPatternSelectorCreator } from "../../../../state/branchPatterns/selectors"
import { BranchPatternInputActionCreator } from "../../../../state/branchPatterns/actions"
import {
  BranchPatternId,
  BranchPatternNumericInputs,
  BranchPatternNumericInputKey,
} from "../../../../state/branchPatterns/types"

interface SliderInputProps<Key extends BranchPatternNumericInputKey> {
  branchPatternId: BranchPatternId
  selectorCreator: BranchPatternSelectorCreator<BranchPatternNumericInputs[Key]>
  actionCreator: BranchPatternInputActionCreator<Key>
  label: string
  min: number
  max: number
  step: number
}

const SliderInput = <Key extends BranchPatternNumericInputKey>({
  branchPatternId,
  selectorCreator,
  actionCreator,
  label,
  min,
  max,
  step,
}: SliderInputProps<Key>) => {
  const value = useSelector(selectorCreator(branchPatternId))
  const dispatch = useDispatch()
  const domId = `${branchPatternId}/${label.toLowerCase()}`
  return (
    <>
      <label htmlFor={domId}>{label}</label>
      <input
        type="range"
        value={value}
        id={domId}
        min={min}
        max={max}
        step={step}
        onChange={(event) =>
          dispatch(actionCreator(branchPatternId, Number(event.target.value)))
        }
      />
    </>
  )
}

export default SliderInput
