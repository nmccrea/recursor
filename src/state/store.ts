/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { configureStore, combineReducers } from "@reduxjs/toolkit"
import similarities from "./similarities/reducer"

const rootReducer = combineReducers({ similarities })
const createStore = () => configureStore({ reducer: rootReducer })

export { createStore }
export type RootState = ReturnType<typeof rootReducer>
export type Selector<ReturnType> = (state: RootState) => ReturnType | undefined
