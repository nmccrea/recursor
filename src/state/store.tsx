/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { combineReducers, configureStore } from "@reduxjs/toolkit"

const rootReducer = combineReducers({})
export type RootState = ReturnType<typeof rootReducer>
export default configureStore({ reducer: rootReducer })
