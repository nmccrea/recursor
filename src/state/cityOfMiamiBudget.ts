/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { RootState } from "./store"
import {
  createAsyncThunk,
  createSlice,
  CaseReducer,
  PayloadAction,
} from "@reduxjs/toolkit"

// State
enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Case Reducers
const fetchDataPending: CaseReducer<CityOfMiamiBudgetState, PayloadAction> = (
  previousState
) => ({
  ...previousState,
  status: Status.Pending,
})

const fetchDataFulfilled: CaseReducer<
  CityOfMiamiBudgetState,
  PayloadAction<object>
> = (previousState, action) => ({
  ...previousState,
  status: Status.Fulfilled,
  data: action.payload,
})

// Async Thunks
const fetchData = createAsyncThunk("fetchData", async (url: string) => {
  const response = await fetch(url)
  return await response.json()
})

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

// Selectors
const getStatus = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.status) || undefined

const getData = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.data) || undefined

export default reducer
export { fetchData, getStatus, getData, Status }
