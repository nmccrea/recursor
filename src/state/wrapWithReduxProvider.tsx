/**
 * Implements Gatsby's `wrapRootElement` API.
 * @packageDocumentation
 */

import React from "react"
import { Provider as ReduxProvider } from "react-redux"
import store from "./store"

interface Arg {
  element: React.ReactNode
}
export const wrapWithReduxProvider = ({ element }: Arg) => (
  <ReduxProvider store={store}>{element}</ReduxProvider>
)
