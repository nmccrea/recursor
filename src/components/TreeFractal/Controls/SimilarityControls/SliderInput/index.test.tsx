import React from "react"
import "jest-styled-components"
import { Provider } from "react-redux"
import { render, fireEvent, getByRole } from "@testing-library/react"
import { createStore, RootState } from "state/store"
import {
  getScaleSetterFor,
  getTranslationSetterFor,
} from "state/similarities/actions"
import {
  getScaleSelectorFor,
  getTranslationSelectorFor,
} from "state/similarities/selectors"
import SliderInput from "."

describe("`<SliderInput />`", () => {
  describe("rendering", () => {
    it("renders correctly based on the value returned by the selector", () => {
      const state: RootState = {
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 0.1256,
              scale: 0.0353,
              angle: Math.PI / 3,
              depth: 2,
              color: "gray",
            },
          },
        },
      }

      const { container } = render(
        <Provider store={createStore(state)}>
          <SliderInput
            label="Test Subject - Translation"
            htmlId="test-subject-translation"
            min={0.824}
            max={0.562}
            step={0.001}
            selector={getTranslationSelectorFor("test/subject")}
            actionCreator={getTranslationSetterFor("test/subject")}
            valueFormatter={jest.fn((value) => String(value))}
            unit="x"
          />
          <SliderInput
            label="Test Subject - Scale"
            htmlId="test-subject-scale"
            min={0.422}
            max={0.744}
            step={0.001}
            selector={getScaleSelectorFor("test/subject")}
            actionCreator={getScaleSetterFor("test/subject")}
            valueFormatter={jest.fn((value) => String(value))}
            unit="rad"
          />
        </Provider>
      )

      expect(container).toMatchSnapshot()
    })

    it("formats the value using to the given formatter", () => {
      const state: RootState = {
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 0.6732,
              scale: 0.8592,
              angle: -Math.PI / 3,
              depth: 3,
              color: "violet",
            },
          },
        },
      }

      const { container } = render(
        <Provider store={createStore(state)}>
          <SliderInput
            label="Test Input"
            htmlId="test-input"
            min={0}
            max={10}
            step={0.1}
            selector={getTranslationSelectorFor("test/subject")}
            actionCreator={getTranslationSetterFor("test/subject")}
            valueFormatter={() => "fake formatted value"}
            unit="x"
          />
        </Provider>
      )

      expect(container).toMatchSnapshot()
    })

    it("converts the output value using the provided output converter", () => {
      const state: RootState = {
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 555,
              scale: 1,
              angle: 0,
              depth: 1,
              color: "violet",
            },
          },
        },
      }
      const valueConverters = {
        convertInput: (value: number) => value * 100,
        convertOutput: (value: number) => value * 0.01,
      }

      const { container } = render(
        <Provider store={createStore(state)}>
          <SliderInput
            label="Test Input"
            htmlId="test-input"
            min={0}
            max={10}
            step={0.1}
            selector={getTranslationSelectorFor("test/subject")}
            actionCreator={getTranslationSetterFor("test/subject")}
            valueFormatter={jest.fn((value) => String(value))}
            unit="x"
            valueConverters={valueConverters}
          />
        </Provider>
      )

      expect(container).toMatchSnapshot()
    })
  })

  describe("change event", () => {
    it("responds correctly to a change event", () => {
      const store = createStore({
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 5.5,
              scale: 1,
              angle: 0,
              depth: 1,
              color: "gray",
            },
          },
        },
      })
      const { container } = render(
        <Provider store={store}>
          <SliderInput
            label="Test Input"
            htmlId="test-input"
            min={0}
            max={10}
            step={0.1}
            selector={getTranslationSelectorFor("test/subject")}
            actionCreator={getTranslationSetterFor("test/subject")}
            valueFormatter={jest.fn((value) => String(value))}
            unit="x"
          />
        </Provider>
      )

      fireEvent.change(getByRole(container, "slider"), {
        target: { value: 2.2 },
      })

      expect(store.getState()).toEqual({
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 2.2,
              scale: 1,
              angle: 0,
              depth: 1,
              color: "gray",
            },
          },
        },
      })
    })

    it("converts the input value using the provided input converter", () => {
      const store = createStore({
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 555,
              scale: 1,
              angle: 0,
              depth: 1,
              color: "violet",
            },
          },
        },
      })
      const valueConverters = {
        convertInput: (value: number) => value * 100,
        convertOutput: (value: number) => value * 0.01,
      }
      const { container } = render(
        <Provider store={store}>
          <SliderInput
            label="Test Input"
            htmlId="test-input"
            min={0}
            max={10}
            step={0.1}
            selector={getTranslationSelectorFor("test/subject")}
            actionCreator={getTranslationSetterFor("test/subject")}
            valueFormatter={jest.fn((value) => String(value))}
            unit="x"
            valueConverters={valueConverters}
          />
        </Provider>
      )

      fireEvent.change(getByRole(container, "slider"), {
        target: { value: 7.77 },
      })

      expect(store.getState()).toEqual({
        similarities: {
          ids: ["test/subject"],
          entities: {
            "test/subject": {
              id: "test/subject",
              translation: 777,
              scale: 1,
              angle: 0,
              depth: 1,
              color: "violet",
            },
          },
        },
      })
    })
  })
})
