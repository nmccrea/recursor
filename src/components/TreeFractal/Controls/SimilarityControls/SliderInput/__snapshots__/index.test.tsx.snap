// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<SliderInput />\` rendering converts the output value using the provided output converter 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(3,minmax(0,1fr));
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c1 {
  grid-column: span 3 / span 3;
  font-size: 0.9em;
}

.c2 {
  grid-column: span 2 / span 2;
}

.c3 {
  margin-left: 0.25rem;
  font-size: 0.9em;
  justify-self: flex-end;
  white-space: nowrap;
}

.c5 {
  justify-self: flex-start;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c6 {
  justify-self: flex-end;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c4 {
  font-size: 0.8em;
  font-style: italic;
  color: gray;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test-input"
    >
      Test Input
    </label>
    <input
      class="c2"
      id="test-input"
      max="10"
      min="0"
      step="0.1"
      type="range"
      value="5.55"
    />
    <span
      class="c3"
    >
      <output
        for="test-input"
      >
        5.55
      </output>
      <span
        class="c4"
      >
         
        x
      </span>
    </span>
    <span
      class="c5"
    >
      0
    </span>
    <span
      class="c6"
    >
      10
    </span>
  </div>
</div>
`;

exports[`\`<SliderInput />\` rendering formats the value using to the given formatter 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(3,minmax(0,1fr));
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c1 {
  grid-column: span 3 / span 3;
  font-size: 0.9em;
}

.c2 {
  grid-column: span 2 / span 2;
}

.c3 {
  margin-left: 0.25rem;
  font-size: 0.9em;
  justify-self: flex-end;
  white-space: nowrap;
}

.c5 {
  justify-self: flex-start;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c6 {
  justify-self: flex-end;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c4 {
  font-size: 0.8em;
  font-style: italic;
  color: gray;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test-input"
    >
      Test Input
    </label>
    <input
      class="c2"
      id="test-input"
      max="10"
      min="0"
      step="0.1"
      type="range"
      value="0.6732"
    />
    <span
      class="c3"
    >
      <output
        for="test-input"
      >
        fake formatted value
      </output>
      <span
        class="c4"
      >
         
        x
      </span>
    </span>
    <span
      class="c5"
    >
      fake formatted value
    </span>
    <span
      class="c6"
    >
      fake formatted value
    </span>
  </div>
</div>
`;

exports[`\`<SliderInput />\` rendering renders correctly based on the value returned by the selector 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(3,minmax(0,1fr));
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c1 {
  grid-column: span 3 / span 3;
  font-size: 0.9em;
}

.c2 {
  grid-column: span 2 / span 2;
}

.c3 {
  margin-left: 0.25rem;
  font-size: 0.9em;
  justify-self: flex-end;
  white-space: nowrap;
}

.c5 {
  justify-self: flex-start;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c6 {
  justify-self: flex-end;
  font-size: 0.7em;
  color: gray;
  white-space: nowrap;
}

.c4 {
  font-size: 0.8em;
  font-style: italic;
  color: gray;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test-subject-translation"
    >
      Test Subject - Translation
    </label>
    <input
      class="c2"
      id="test-subject-translation"
      max="0.562"
      min="0.824"
      step="0.001"
      type="range"
      value="0.1256"
    />
    <span
      class="c3"
    >
      <output
        for="test-subject-translation"
      >
        0.1256
      </output>
      <span
        class="c4"
      >
         
        x
      </span>
    </span>
    <span
      class="c5"
    >
      0.824
    </span>
    <span
      class="c6"
    >
      0.562
    </span>
  </div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test-subject-scale"
    >
      Test Subject - Scale
    </label>
    <input
      class="c2"
      id="test-subject-scale"
      max="0.744"
      min="0.422"
      step="0.001"
      type="range"
      value="0.0353"
    />
    <span
      class="c3"
    >
      <output
        for="test-subject-scale"
      >
        0.0353
      </output>
      <span
        class="c4"
      >
         
        rad
      </span>
    </span>
    <span
      class="c5"
    >
      0.422
    </span>
    <span
      class="c6"
    >
      0.744
    </span>
  </div>
</div>
`;
